records/2022-12-13.upgrade-database.md

```diff
@@ -0,0 +1,104 @@
+[production](https://ca-central-1.console.aws.amazon.com/rds/home?region=ca-central-1#database:id=notification-canada-ca-production-cluster;is-cluster=true;tab=configuration) are both running aurora-postgresql 11.13 (from 2021-08-12). The latest version of PostgeSQL available on AWS is 14.5. The latest version of PostgeSQL is 15.1.
+
+AWS recommended that we upgrade to at least 11.16 by July 23, 2022.
+
+## Options
+
+According to [Upgrading the PostgreSQL DB engine for Amazon RDS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.PostgreSQL.html) we cannot go straight to 14.5. Our options are:
+- 11.17
+- 12, ie 12.8, 12.9, 12.10, 12.11, or 12.12
+- 13.14 (AWS maximum recommended upgrade)
+
+According to [AWS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.PostgreSQL.html):
+
+**Major version upgrades**
+
+Major version upgrades can contain database changes that are not backward-compatible with existing applications. As a result, you must manually perform major version upgrades of your DB instances. You can initiate a major version upgrade by modifying your DB instance. However, before you perform a major version upgrade, we recommend that you follow the steps described in [Choosing a major version upgrade for PostgreSQL](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.PostgreSQL.html#USER_UpgradeDBInstance.PostgreSQL.MajorVersion). During a major version upgrade, Amazon RDS also upgrades all of your in-Region read replicas along with the primary DB instance.
+
+**Minor version upgrades**
+
+In contrast, minor version upgrades include only changes that are backward-compatible with existing applications. You can initiate a minor version upgrade manually by modifying your DB instance. Or you can enable the Auto minor version upgrade option when creating or modifying a DB instance. Doing so means that your DB instance is automatically upgraded after Amazon RDS tests and approves the new version. If your PostgreSQL DB instance is using read replicas, you must first upgrade all of the read replicas before upgrading the primary instance. If your DB instance is in a Multi-AZ deployment, then the writer and any standby replicas are upgraded simultaneously. Therefore, your DB instance might not be available until the upgrade is complete. For more details, see [Automatic minor version upgrades for PostgreSQL](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.PostgreSQL.html#USER_UpgradeDBInstance.PostgreSQL.Minor). For information about manually performing a minor version upgrade, see [Manually upgrading the engine version](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.Upgrading.html#USER_UpgradeDBInstance.Upgrading.Manual).
+
+### 11.17
+
+Likely not risky.
+
+### 12
+
+There are [breaking changes](https://www.postgresql.org/docs/12/release-12.html#id-1.11.6.18.4) from 11 to 12. Unknown if any affect us.
```

Comment:
Can we use this ADR to investigate the breaking changes and see if that could affect us? 

We could run a local database the target versions and see if that is problematic. Which on that note, I always run the latest locally... so I know some things work fine with GCNotify.